File: blog/notes/JS/group.js
function keyStringValues(row: any, keys: any[]): string {
  return keys.map(x => row[x]).join(', ');
}

function groupAndSum(rows, groupKeys, sumKeys) {
  const res = {};
  rows.forEach(row => {
    key = keyStringValues(row, groupKeys);
    if (res[key]) {
      sumKeys.forEach(sumKey => {
        res[key][sumKey] += row[sumKey] || 0;
      });
    } else {
      res[key] = {};
      groupKeys.forEach(groupKey => {
        res[key][groupKey] = row[groupKey];
      });
      sumKeys.forEach(sumKey => {
        res[key][sumKey] = row[sumKey] || 0;
      });
    }
  });
  return Object.values(res);
}

const rows = [
 {paramId: 'test1', production: 4, losses: 2, count: 1},
 {paramId: 'test1', production: 6, losses: 3, count: 1},
 {paramId: 'test2', production: 1, losses: 0, count: 1},
 {paramId: 'test2', production: 2, losses: 1, count: 1},
 {paramId: 'test2', production: 3, losses: 1, count: 1},
];
console.log(groupAndSum(rows, ['paramId'], ['production', 'losses', 'count']));
/*
[
  {paramId: "test1", production: 10, losses: 5, count: 2},
  {paramId: "test2", production: 6, losses: 2, count: 3}
]
*/

// groupBy ver. 1
function groupBy(rows: any[], keys: any[]): any {
  const groups = {};
  rows.forEach(row => {
    const group = groupStringValues(row, keys);
    groups[group] = groups[group] || [];
    groups[group].push(row);
  });
  return groups;
}

// groupBy ver. 2
export function groupBy<T = any>(rows: T[], keyGetter: (item: T) => any): Map<any, T[]> {
  const map = new Map();
  rows.forEach((item) => {
    const key = keyGetter(item);
    if (!map.has(key)) {
      map.set(key, [item]);
    } else {
      map.get(key).push(item);
    }
  });
  return map;
}
groupBy<MeasureType>(measureTypes, x => x.chartGroupId);


function findMinMax(rows, field) {
  if (rows.length === 0) {
    return [null, null];
  }
  let min = rows[0][field];
  let max = min;
  for (let i = 1, len = rows.length; i < len; i++) {
    const v = field ? rows[i][field] : rows[i];
    min = (v < min) ? v : min;
    max = (v > max) ? v : max;
  }
  return [min, max];
}

File: blog/notes/Python/group.py
from typing import List, Dict


def key_string_values(row: dict, keys: List[str]) -> str:
    return ','.join(map(lambda x: str(row[x]), keys))


def group_and_sum(rows: list, group_keys: List[str], sum_keys: List[str]) -> list:
    res = {}
    for row in rows:
        key = key_string_values(row, group_keys)
        if key in res and res[key] is not None:
            for sum_key in sum_keys:
                res[key][sum_key] += row[sum_key] if row[sum_key] is not None else 0
        else:
            res[key] = {}
            for group_key in group_keys:
                res[key][group_key] = row[group_key]
            for sum_key in sum_keys:
                res[key][sum_key] = row[sum_key] if row[sum_key] is not None else 0

    return list(res.values())


def group_by(rows: List[dict], group_keys: List[str]) -> Dict[str, list]:
    res = {}
    for row in rows:
        key = key_string_values(row, group_keys)
        if key in res and res[key] is not None:
            res[key].append(row)
        else:
            res[key] = [row]
    return res


def get_min_max(rows: list, field: str) -> Tuple[Optional[float], Optional[float]]:
    if len(rows) == 0:
        return None, None
    field_values = [x[field] for x in rows]
    if all(x is None for x in field_values):
        return None, None
    return min(field_values), max(field_values)


def get_last_group_rows(rows: List[dict], group_keys: List[str], sort_key: Callable, count: int) -> list:
    """ Последние n строк для каждой группы """
    res = []
    count = int(count)
    grouped = group_by(rows, group_keys)
    for key in grouped:
        param_rows = grouped[key]
        if len(param_rows) > count:
            sorted_rows = sorted(param_rows, key=sort_key)
            res = res + sorted_rows[-count:]
        else:
            res = res + param_rows
    return res
